event.cpp

```diff
@@ -38,21 +38,30 @@ TCHAR *message_string(unsigned long error) {
 /* Log a message to the Event Log */
 void log_event(unsigned short type, unsigned long id, ...) {
   va_list arg;
-  int count;
-  TCHAR *s;
   TCHAR *strings[NSSM_NUM_EVENT_STRINGS];
+  TCHAR *buffer = NULL;
+  int len;
+
+  TCHAR *format = message_string(id);
+  if (!format) return;
 
   /* Open event log */
   HANDLE handle = RegisterEventSource(0, NSSM_SOURCE);
   if (! handle) return;
 
   /* Log it */
-  count = 0;
   va_start(arg, id);
-  while ((s = va_arg(arg, TCHAR *)) && count < NSSM_NUM_EVENT_STRINGS - 1) strings[count++] = s;
-  strings[count] = 0;
-  va_end(arg);
-  ReportEvent(handle, type, 0, id, 0, count, 0, (const TCHAR **) strings, 0);
+  len = _vsctprintf_p(format, arg) + 1;
+  buffer = (TCHAR*)malloc(len * sizeof(TCHAR));
+  if (buffer)
+  {
+    _vstprintf_p(buffer, len, format, arg);
+    strings[0] = buffer;
+    strings[1] = 0;
+    va_end(arg);
+    ReportEvent(handle, type, 0, type, 0, 1, 0, (const TCHAR **) strings, 0);

```

Comment:
```suggestion
    ReportEvent(handle, type, 0, id, 0, 1, 0, (const TCHAR **) strings, 0);
```

Reply:
we can't use id anymore as EventCreate.exe does not know about message resource of nssm

Reply:
Interesting that sending type as id works :)

Reply:
type is error level, and is already defined in EventCreate, though this is not the best way to use it.
https://docs.microsoft.com/en-gb/windows/win32/api/winbase/nf-winbase-reporteventa?redirectedfrom=MSDN